### DISCUSSION TITLE
[CX16] Where can I find more information (Example code) on how Banking is done on the cx16

### DISCUSSION THREAD

**Discussion by RetrofanC** [cc65/cc65#1595] [Q&A] [answered]
Hi, 

Going through the different configuration files the cx16-banking.cfg looks particular interesting. 
It might give some further ideation or aspects I might be able to re-use. 

Hence, where could I find some more information on how this is intended to work and is there a contact who has been working on this who could be asked some questions. 

Just as a heads up some of my questions are; 

[Q1] Every bank is sent to a separate file and how does these files are being used? 
Are they loaded in like a 'overlay'  in a similar way as on the c128 ?
Alternatively, are they all 'pre-loaded' at the start and is there some real 'banking switching' happening between the banks? 

[Q2] As the bank=$xx option is being used in the *.cfg file how is this being using in the actual code? 
Is there example code available how the banks are being determined and used in the code? 
(I know the examples from the manual but would benefit from seeing some actual code. 

[Q3] How are the banks being used in the actual C code. So how is indicated that a certain "void function(void)"  is using a certain bank. (is this done with a #pragma in the C-Code or with SEGMENTS in the *.s perhaps? 

[Q4] While running the C-Code how does the C-Code makes use of a kind of "look-up table" to find the right bank belong to a function. Consequentially, how is the code switching banks to make sure the call to the function is done to the right bank. Furthermore, how does the routine works which makes sure that the function runs in the right memory/bank and returns to the right memory as well to continue further execution of the main program. 

Again the above just to learn and re-use the experience and work of others who already made this kind of function work. 
As my end goals just a getting a working solution as soon (and smooth) as possible which can be used by the community.  

Thanks, 

cx16-banking.cfg

FEATURES {
    STARTADDRESS: default = $0801;
}
SYMBOLS {
    __LOADADDR__:     type = import;
    __EXEHDR__:       type = import;
    __BANKRAMADDR__:  type = import;
    __STACKSIZE__:    type = weak,   value = $0800; # 2K stack
    __HIMEM__:        type = weak,   value = $9F00;
    __BANKRAMSTART__: type = export, value = $A000;
    __BANKRAMSIZE__:  type = weak,   value = $2000; # 8K banked RAM
}
MEMORY {
    ZP:         file = "", define = yes,  start = $0022,                size = $0080 - $0022;
    LOADADDR:   file = %O,                start = %S - 2,               size = $0002;
    HEADER:     file = %O, define = yes,  start = %S,                   size = $000D;
    MAIN:       file = %O, define = yes,  start = __HEADER_LAST__,      size = __HIMEM__ - __HEADER_LAST__;
    BSS:        file = "",                start = __ONCE_RUN__,         size = __HIMEM__ - __ONCE_RUN__ - __STACKSIZE__;
    BRAM01ADDR: file = "%O.01",           start = __BANKRAMSTART__ - 2, size = $0002;
    BRAM01:     file = "%O.01",           start = __BANKRAMSTART__,     size = __BANKRAMSIZE__, bank = $01;
    BRAM02ADDR: file = "%O.02",           start = __BANKRAMSTART__ - 2, size = $0002;
    BRAM02:     file = "%O.02",           start = __BANKRAMSTART__,     size = __BANKRAMSIZE__, bank = $02;
    BRAM03ADDR: file = "%O.03",           start = __BANKRAMSTART__ - 2, size = $0002;
    BRAM03:     file = "%O.03",           start = __BANKRAMSTART__,     size = __BANKRAMSIZE__, bank = $03;
    BRAM04ADDR: file = "%O.04",           start = __BANKRAMSTART__ - 2, size = $0002;

**Comment by oliverschmidt** [comment, marked as answer]
Maybe you aren't aware that from the linker perspective 'banking' is equal to 'overlays'. In both scenarios, the same address is used by different code at different times. Therefore https://github.com/cc65/cc65/blob/master/samples/overlaydemo.c says all there is to say about using banking from/with C code.

**Reply by RetrofanC** [reply]
Thanks Oliver, for the clarification most helpful again to understand the cc65 tool chain and it's language/terminology used better. 

Understood, that the overlay.c (again which is a very nice piece of code I my eyes) is the source for the mechanism of banking/overlaying.
